### DISCUSSION TITLE
클로저 함수 관련 질문입니다.

### DISCUSSION THREAD

**Discussion by ghost** [tastekim/WeAllLie-BE#42] [Q&A] [answered]
코드가 한결 간결해질 것 같아서 저도 _id, nickname에 붙는 숫자에 적용하려고 했는데요,
서버가 켜져있는 상태에서는 잘 동작하는데 서버를 끄면 번호가 초기화가 되네요..

다른 함수들의 리팩토링을 생각해서 클로저 함수를 클래스 내의 하나의 메소드로 만들어서 파일을 따로 만들었는데요
쉬운 방법은 클로저 함수의 초기값 a 를 현재 디비의 유저 다큐먼트의 마지막 _id 번호를 다시 가져오는 방법이 있긴 하지만...
결국 db를 한 번은 거치게 되어야 하네요.
그렇더라도 어짜피 기존 코드도 db를 거치도록 되어 있엇는데 코드가 훨씬 간결해져서 좋긴해요!!! 

아니면 매번 변경되는 autoNum의 값을 다시 클로저 함수의 초기값으로 변경지정이 될까 싶긴 한데 될지 모르겠고..

혹시 db를 거치지 않는 다른 방법이 있을까요?

**Comment by tastekim** [comment, marked as answer]
클로저 함수 내의 변수는 내부함수가 참조하는 상태고 저희 코드상 autoInc라는 변수에 autoIncrease의 return값인 inner를 담겨있어 autoIncrease 함수 안에 있는 변수 a를 계속 참조하고 있기 때문에 함수 컨텍스트가 종료되어도 내부적으로 변수 a의 참조컨텍스트가 0이 아니여서 메모리상에 계속 살아있는거고 그런 특징을 이용해서 함수를 실행시킬때마다 1씩 증가시켜서 사용합니다. 다만 저 a도 전역컨텍스트가 종료되는 시점까지 살아있게 되기 때문에 저희 상황에서는 서버가 종료되는 시점에서야 비로소 a는 다시 1이 되기 때문에 서버가 종료된다면 서비스를 이용하는 유저도 없고 생성되있는 방이 있을수 없기 때문에 그 부분을 이용해서 클로저함수의 특징을 이용해서 autoNum을 사용했습니다 !

민형님은 서버가 종료되더라도 유저의 정보는 그대로 유지되어야하는 상황이여서 아무래도 클로저함수보단 DB를 통해 데이터를 확인 후에 Int값을 부여해야할것 같긴 하네요,,,

**Reply by tastekim** [reply]
클로저가 '내부함수가 외부함수 컨텍스트를 참조할 때 생기는 특별한 현상' 인데 이 현상은 전역 컨텍스트가 종료되는 시점(서버가 종료되는 시점)에 다시 초기화 되어버립니다ㅜ
